tests/http/test_webclient.py

```diff
@@ -0,0 +1,106 @@
+from synapse.http.site import SynapseSite
+
+from tests.server import make_request
+from tests.unittest import HomeserverTestCase, create_resource_tree, override_config
+
+
+class WebClientTests(HomeserverTestCase):
+    @override_config(
+        {
+            "web_client_location": "https://example.org",
+        }
+    )
+    def test_webclient_resolves_with_client_resource(self):
+        """
+        Tests that both client and webclient resources can be accessed simultaneously.
+
+        This is a regression test created in response to https://github.com/matrix-org/synapse/issues/11763.
+        """
+        for resource_name_order_list in [
+            ["webclient", "client"],
+            ["client", "webclient"],
+        ]:
+            # Create a dictionary from path regex -> resource
+            resource_dict: Dict[str, Resource] = {}
+
+            for resource_name in resource_name_order_list:
+                resource_dict.update(
+                    SynapseHomeServer._configure_named_resource(self.hs, resource_name)
+                )
+
+            # Create a root resource which ties the above resources together into one
+            root_resource = Resource()
+            create_resource_tree(resource_dict, root_resource)
+
+            # Create a site configured with this resource to make HTTP requests against
+            listener_config = ListenerConfig(
+                port=8008,
+                bind_addresses=["127.0.0.1"],
+                type="http",
+                http_options=HttpListenerConfig(
+                    resources=[HttpResourceConfig(names=resource_name_order_list)]
+                ),
+            )
+            test_site = SynapseSite(
+                logger_name="synapse.access.http.fake",
+                site_tag=self.hs.config.server.server_name,
+                config=listener_config,
+                resource=root_resource,
+                server_version_string="1",
+                max_request_body_size=1234,
+                reactor=self.reactor,
+            )
+
+            # Attempt to make requests to endpoints on both the webclient and client resources
+            # on test_site.
+            self._request_client_and_webclient_resources(test_site)
+
+    def _request_client_and_webclient_resources(self, test_site: SynapseSite) -> None:
+        """Make a request to an endpoint on both the webclient and client-server resources
+        of the given SynapseSite.
+
+        Args:
+            test_site: The SynapseSite object to make requests against.
+        """
+
+        # Ensure that the *webclient* resource is behaving as expected (we get redirected to
+        # the configured web_client_location)
+        channel = make_request(
+            self.reactor,
+            site=test_site,
+            method="GET",
+            path="/_matrix/client",
+        )
+        self.assertEqual(channel.code, 302)
```

Comment:
Could we have an `HTTPStatus....` here? I can never remember what anything other than `200, 400, 404, 500` means. For consistency, ditto for the 200 below.

Reply:
fwiw most tests use integers for status codes (and so does most of the servlet code when specifying the status to respond with). Though a few bits of the code use `HTTPStatus`, so it looks like we're already inconsistent on this, so :shrug:

Reply:
I initially did this as I like to use literals in tests where possible (to ensure what we have in the code makes sense). However if these status codes ever change (unlikely) then presumably we'll want to update the tests anyhow.

Reply:
I updated the 302 to an `HTTPStatus.FOUND`. The status code name is pretty vague, so I added a comment on top as well.

Reply:
> fwiw most tests use integers for status codes (and so does most of the servlet code when specifying the status to respond with). Though a few bits of the code use `HTTPStatus`, so it looks like we're already inconsistent on this, so 🤷

I think we've been slowly updating to use `HTTPStatus`. That seems to be @DMRobertson's plan at least.

Reply:
That's right; maybe I ought to just do it in one big bang if we're going to use them at all.

Being fully candid, I don't have a justification for this except for "I think it's nicer". Happy to drop this if people find it noisey or not useful.

Reply:
I think it's nicer and more useful than the status code alone. And if one wants to know the status code, they can simply check the definition.